src/main/scala/com/microsoft/hyperspace/index/rankers/JoinIndexRanker.scala

```diff
@@ -33,16 +33,28 @@ object JoinIndexRanker {
    * number of buckets, better the parallelism achieved during join in general, assuming there is
    * no resource constraint.
    *
+   * If hybridScanEnabled is true, pick the index pair which covers the most source data
+   * so that we could minimize the amount of data for on-the-fly shuffle or merge.
+   *
    * @param indexPairs index pairs for left and right side of the join. All index pairs are
    *                   compatible with each other.
+   * @param hybridScanEnabled HybridScan config.
    * @return rearranged index pairs according to their ranking. The first is the best.
    */
   def rank(
-      indexPairs: Seq[(IndexLogEntry, IndexLogEntry)]): Seq[(IndexLogEntry, IndexLogEntry)] = {
+      indexPairs: Seq[(IndexLogEntry, IndexLogEntry)],
+      hybridScanEnabled: Boolean = false): Seq[(IndexLogEntry, IndexLogEntry)] = {
 
     indexPairs.sortWith {
       case ((left1, left2), (right1, right2)) =>
-        if (left1.numBuckets == left2.numBuckets && right1.numBuckets == right2.numBuckets) {
+        if (hybridScanEnabled && ((left1.allSourceFileInfos.size + left2.allSourceFileInfos.size) >
```

Comment:
Could you please confirm if you intend to use number of files or sum of size of all files here?
`left1.allSourceFileInfos.size` vs `left1.allSourceFileInfos.map(_.size).reduce(_ + _)`

The first one only gives a count of files, not the size of data being indexed.

Reply:
I only considered "append-only" dataset for this, so the number of files is enough to check their total size.
To support delete & append, yes we need to calculate the summation of each candidate index.
We might need to keep the value somewhere rather than calculating the size every time.

Reply:
got it. thanks.